lib/statistex.ex

```diff
@@ -401,13 +450,44 @@ defmodule Statistex do
   defdelegate(percentiles(samples, percentiles), to: Percentile)
 
   @doc """
-  Calculates the mode of the given samples.
+  A map showing which sample occurs how often in the samples.
 
+  Goes from a concrete occurence of the sample to the number of times it was observed in the samples.
 
-  Mode is the sample(s) that occur the most. Often one value, but can be multiple values if they occur the same amount of times. If no value occurs at least twice, this value will be nil.
+  ## Examples
+
+      iex> Statistex.frequency_distribution([1, 2, 4.23, 7, 2, 99])
+      %{
+        2 => 2,
+        1 => 1,
+        4.23 => 1,
+        7 => 1,
+        99 => 1
+      }
+
+      iex> Statistex.frequency_distribution([])
+      ** (ArgumentError) Passed an empty list ([]) to calculate statistics from, please pass a list containing at least on number.
+  """
+  @spec frequency_distribution(samples) :: %{required(sample) => pos_integer}
+  def frequency_distribution([]), do: raise(ArgumentError, @empty_list_error_message)
+
+  def frequency_distribution(samples) do
+    Enum.reduce(samples, %{}, fn sample, counts ->
+      Map.update(counts, sample, 1, fn old_value -> old_value + 1 end)
```

Comment:
:nail_care:  suggestion : This style is not exactly functional. Consider writing a private function based on recursion, like so:
https://elixir-lang.org/getting-started/recursion.html

Reply:
Also, it's important to point out that, for large samples, Enum will become very slow. One should consider using `Stream` or even, for very huge data, GenStage, something like `Flow`: https://hexdocs.pm/flow/Flow.html
Bear in mind that erlang was not built for mathematical computation, so it won't have a performance as good as other languages (python or java for example). It was built with communication, parallel processing and scalability.
Take a look here: https://stackoverflow.com/questions/13629142/why-is-erlang-slower-than-java-on-all-these-small-math-benchmarks#13629562

Reply:
Hi there,

thanks for your input.

re functional: For me FP is more about separation of side effects/pure functions and maybe immutable data structures. I'm well familiar with the concept of recursion and also use it in this library. I know lots of FP code uses recursion, reduce is also arguably very functional. If you have a recursive implementation of this function that is more readable and/or faster I'm happy to accept PRs.

re performance: I'm also aware of the performance trade offs of Erlang, please refer to the [performance section of the README](https://github.com/bencheeorg/statistex#performance). Parallelization is currently not really a focus of Statistex but might be an option later. The major client of statistex (benchee) already calculates the statistics in parallel for multiple data sets so this would be unlikely to yield performance improvements.